src/vm/dllimportcallback.cpp

```diff
@@ -33,6 +33,90 @@ struct UM2MThunk_Args
     int argLen;
 };
 
+class UMEntryThunkFreeList
+{
+public:
+    UMEntryThunkFreeList(size_t size) :
+        m_size(size),
+        m_count(0),
+        m_crst(CrstUMEntryThunkFreeList)
+    {
+        WRAPPER_NO_CONTRACT;
+    }
+
+    UMEntryThunk *GetUMEntryThunk()
+    {
+        WRAPPER_NO_CONTRACT;
+
+        if (m_count < m_size)
+            return NULL;
+
+        CrstHolder ch(&m_crst);
+
+        SListElem<UMEntryThunk*> *pElem = m_list.RemoveHead();
+
+        if (pElem == NULL)
+            return NULL;
+
+        --m_count;
+
+        return pElem->GetValue();
+    }
+
+    void AddToList(UMEntryThunk *pThunk)
+    {
+        WRAPPER_NO_CONTRACT;
+
+        CrstHolder ch(&m_crst);
+
+        if (m_count == m_size)
+        {
+            SListElem<UMEntryThunk*> *pElem = m_list.RemoveHead();
+
+            if (pElem != NULL)
+            {
+                UMEntryThunkFreeList::FreeThunk(pElem->GetValue());
```

Comment:
We can put the thunk to the list all the time. No need to ever return it back to the LoaderHeap.

Reply:
If you do this, we can also delete `LHF_ZEROINIT` flag on the LoaderHeap added by previous iteration of this change.

Reply:
Why we don't need to return allocated chunks to the LoaderHeap? I think they can be reused since GlobalLoaderAllocator's executable heap is also used in other places.
Thank you!

Reply:
Outside UMEntryThunks, the executable heap is used in a very few rarely used places. There is a close to zero chance that the returned memory would be reused for anything but UMEntryThunks.

Reply:
Thank you for explanation, I've removed returning memory to the LoaderHeap in `UMEntryThunkFreeList`.
Is `LHF_ZEROINIT` flag not useful? I think it can reduce time of allocation if we don't need initialized memory.

Reply:
LoaderHeap allocates memory directly from the OS using mmap. This memory is zero initialized, so the zero initialization is free for the normal loader head use.

ZeroInit can only save something for cases where the memory is returned back to LoaderHeap. This was only done by UMEntryThunks on mainline paths. After this change, it will pretty much never happen. The memory is generally returned to LoaderHeap on exceptional paths only, like a complex operations like type loading fails in the middle and we need to backout the memory allocated so far - so that we do not have a memory leak if this operation is repeated again and again. We do not optimize performance on exceptional paths. We prefer simplicity for exceptional paths.

Reply:
Thank you! I've removed this option.